docs/decisions/0009-baseline-error-handling-for-consumer.rst

```diff
@@ -0,0 +1,48 @@
+
+.. _Confluent Kafka error codes: https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#kafkaerror
+
+Additionally, the consumer has a limited number of options when it encounters an error. If it cannot retrieve an event, the best it can do is keep trying. If it cannot parse or identify a received event, it can either halt (wait for manual intervention) or start dropping events. And if a signal receiver is in error, the consumer can at best note this somehow.
+
+Halting and waiting for intervention would be appropriate when the consumer is *completely broken*, but would be inappropriate when only a small subset of events cannot be processed. We may not be able to write code that can tell the difference between these situations. Depending on the deployment infrastructure, allowing an exception to propagate out of the consumer loop might result in the consumer being restarted (no overall effect), or it might shut down the consumer entirely until manually restarted (creating a halt).
+
+Decision
+********
+
+We will start by catching *every* raised exception before it can escape the consumer loop, logging it with as much information as we think will be necessary for debugging or for restarting the consumer. There's currently no machinery for restarting a consumer group at a particular offset, but it would allow operators to recover from various types of failures (transient, data-specific, and buggy consumer) if needed.
+
+Consumers of the Event Bus are generally expected to tolerate duplicate events (at-least-once delivery) and we know that some of the planned applications are predicated on low latency. Since topics can always be replayed to consumers, failing fast and moving on to the next event is preferable.
+
+As we gain experience, we can add special handling for certain known cases that would be better served by a backoff, retry, or halt.
+
+Receiver exceptions are not *raised*, per se, but are returned by ``send_robust``. We will log these but not treat these as consumer errors, as these exceptions are much more likely to be related to IDA business logic than to the Event Bus itself. There may also be multiple receivers, and failures in one do not indicate that other receivers should stop receiving the event stream.
+
+All raised errors will also be sent to telemetry (New Relic) for monitoring.
+
+Consequences
+************
+
+The consumer will err on the side of low latency between IDAs, creating higher (or at least sooner) inter-IDA consistency at the possible expense of inter-object consistency within a topic (if some events are dropped and other processed). We will have to ensure that we capture all failures sufficiently durably that we can replay them, and also that receivers are capable of "healing" their data when a topic is replayed.
+
+Rejected Alternatives
```

Comment:
Not sure whether to call this Deferred/Rejected Alternatives, but it would be great to see something around the following. I think it is especially important to use the term Dead Letter Queue, because it may be an expected piece of functionality, and it would be good to clarify explicitly, and not just implicitly.

* It is a common practice to introduce a new topic as a Dead Letter Queue (DLQ), or a place to send events that we were unable to process. For now, the log will serve as our DLQ, but a topic of messages may be simpler to re-process (although the messages will no longer be ordered). This work has been deferred as of the initial draft of this ADR.
* Additionally, sometimes a topic or topics are introduced for retrying events before the events are moved to the DLQ. Again, there are design considerations required to maintain order. This work has been deferred as of the initial draft of this ADR.

Reply:
Yeah, DLQ was the last thing I felt was missing, but I was still mulling over what to write. Thanks for the suggestions!

Reply:
Nit: Would you mind renaming to "Deferred/Rejected Alternatives"? That would be my preference, given the content, but understood if you feel differently (i.e. are more concerned with consistent headers). I've used my suggestion elsewhere, I think.

Reply:
I'd be OK with renaming it.